### PR TITLE
Handle cast function

### PR DESCRIPTION
**Pull request by soma-mondal**
Handle Cast to Time & Timestamp in BQ, Hive, Spark

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/util/CastCallBuilder.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    this.dialect = dialect;
+  }
+
+  public SqlNode makCastCallForTimestampWithPrecision(SqlNode operandToCast, int precision) {
+    SqlNode timestampWithoutPrecision =
+        dialect.getCastSpec(new BasicSqlType(RelDataTypeSystem.DEFAULT, SqlTypeName.TIMESTAMP));
+    SqlCall castedTimestampNode = CAST.createCall(POS, operandToCast,
+        timestampWithoutPrecision);
+    SqlCharStringLiteral timestampFormat =
+        SqlLiteral.createCharString("YYYY-MM-DDBHH24:MI:SS.S(" + precision + ")",
```
Comment (Bindi1993) [inline core/src/main/java/org/apache/calcite/util/CastCallBuilder.java]: please check idf we can use string.format

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java [hunk 2]
```diff
@@ -173,6 +238,28 @@ public BigQuerySqlDialect(SqlDialect.Context context) {
...
+    }
+    return super.getCastCall(operandToCast, castFrom, castTo);
+  }
+
+  private SqlNode makCastCallForTimeWithPrecision(SqlNode operandToCast, int precision) {
+    SqlParserPos pos = SqlParserPos.ZERO;
+    SqlNode timeWithoutPrecision =
+        getCastSpec(new BasicSqlType(RelDataTypeSystem.DEFAULT, SqlTypeName.TIME));
+    SqlCall castedTimeNode = CAST.createCall(pos, operandToCast, timeWithoutPrecision);
+    SqlCharStringLiteral timeFormat = SqlLiteral.createCharString("HH24:MI:SS.S(" + precision + ")",
```
Comment (Bindi1993) [inline core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java]: Please use String.format for concatenation

File: core/src/main/java/org/apache/calcite/sql/SqlDialect.java [hunk 3]
```diff
@@ -1218,6 +1240,50 @@ public static String replace(
...
+
+  private Pair<List<String>, List<Character>> getDateTimeTokensWithSeparators(
+      String standardDateFormat) {
+    List<String> dateTimeTokens = new ArrayList<>();
+    List<Character> separators = new ArrayList<>();
+    int startIndex = 0;
+    int lastIndex = standardDateFormat.length() - 1;
+    for (int i = 0; i <= lastIndex; i++) {
+      if (i == lastIndex) {
+        dateTimeTokens.add(standardDateFormat.substring(startIndex));
```
Comment (Bindi1993) [inline core/src/main/java/org/apache/calcite/sql/SqlDialect.java]: can we modify the logic? we can apply the loop from 0 to less than lastIndex and separate the if logic from the loop

File: core/src/main/java/org/apache/calcite/sql/SqlDialect.java [hunk 4]
```diff
@@ -1218,6 +1240,50 @@ public static String replace(
...
+    }
+    assert dateTimeTokens.size() - 1 == separators.size();
+    return new Pair<>(dateTimeTokens, separators);
+  }
+
+  private String getFinalFormat(
+      List<String> dateTimeTokens, List<Character> separators,
+      Map<SqlDateTimeFormat, String> dateTimeFormatMap) {
+    StringBuilder finalFormatBuilder = new StringBuilder();
+    for (String token : dateTimeTokens) {
```
Comment (Bindi1993) [inline core/src/main/java/org/apache/calcite/sql/SqlDialect.java]: please check if we can refactor this method

File: core/src/main/java/org/apache/calcite/sql/dialect/HiveSqlDialect.java [hunk 5]
```diff
@@ -132,6 +203,26 @@ public HiveSqlDialect(Context context) {
     }
   }
 
+  @Override public SqlNode getCastCall(
+      SqlNode operandToCast, RelDataType castFrom, RelDataType castTo) {
+    if (castTo.getSqlTypeName() == SqlTypeName.TIMESTAMP && castTo.getPrecision() > 0) {
+      return new CastCallBuilder(this).makCastCallForTimestampWithPrecision(operandToCast,
+          castTo.getPrecision());
```
Comment (Bindi1993) [inline core/src/main/java/org/apache/calcite/sql/dialect/HiveSqlDialect.java]: Please create CastBuilder object once and use reference variable

Comment (soma-mondal) [reply core/src/main/java/org/apache/calcite/sql/dialect/HiveSqlDialect.java]: Done

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java (hunk 2 above)
Comment (soma-mondal) [reply core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java]: Done

File: core/src/main/java/org/apache/calcite/sql/SqlDialect.java (hunk 3 above)
Comment (soma-mondal) [reply core/src/main/java/org/apache/calcite/sql/SqlDialect.java]: done

File: core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java [hunk 6]
```diff
@@ -5299,6 +5299,201 @@ public void testIfMethodArgument() {
     assertThat(toSql(root, DatabaseProduct.HIVE.getDialect()), isLinux(expectedHive));
   }
 
+
+  @Test public void testCastToTimestamp() {
+    String query = "SELECT cast(\"birth_date\" as TIMESTAMP) "
+        + "FROM \"foodmart\".\"employee\"";
+    final String expected = "SELECT CAST(birth_date AS TIMESTAMP(0))\n"
```
[the remark this answers is not in the document]
Comment (soma-mondal) [reply core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java]: Need to override getCastSpec() for all the dialect to generate right query. Already done for BQ in apache calcite.

**Pull request merged**